### ISSUE TITLE
groupby function

### ISSUE THREAD

**Issue by adamlerer** [pytorch/pytorch#20613]
## Pitch

```
groups, values, inverse_map = T.groupby(F)
```

Basically T is an `Nx...` tensor and `F` is a size `N` tensor. This function returns a list `groups` of length equal to the number of unique values in F. Each group contains all the rows of `T` for that unique element. `values` is just a tensor containing the unique values corresponding to each group, and `inverse_map` is a list of tensors telling you where all the rows came from.

## Alternatives

This can be implemented in about 10 lines in pytorch with something like:
```
sorted_vals, order = F.sort(0)
delta = F[1:] - F[:-1]
cutpoints = delta.nonzero()[0].tolist()
res, inverse_map = [], [], torch.zeros(len(cutpoints))
for start, end, i in zip([0] + cutpoints, cutpoints + [len(F)], range(len(cutpoints) + 1)):
    res.append(T[order[start:end]])
    inverse_map.append(order[start:end])
    values[i] = F[start]

return res, values, inverse_map
```

This could be done faster and in a more standard way as part of Pytorch.

## Additional context

I've ended up needing to implement this multiple times across projects. Say for example you have a batch of data but want to run different models based on the label index, or based on the length of the input, etc. You want to subdivide the batch according to that field, and in some cases you need to reconstruct the outputs back into the shape of the original tensor.

cc @colesbury @lerks

**Comment by Diego999** [comment]
Any news on this or faster way of doing it with 1.5.0 ?
@colesbury @lerks @vishwakftw @jerryzh168
